### DISCUSSION TITLE
Construct Form with IntPtr Handle (MonoGame)

### DISCUSSION THREAD

**Discussion by EclipsedButter** [picoe/Eto#2693] [Q&A]
Hi, I'm in the process of migrating an application from Windows.Forms to Eto. The application was built using winforms and monogame, and initialized the main form for the application with the line `Form MainForm = (Form)Form.FromHandle(Window.Handle)`. So the handle for the new Form here is the IntPtr to an SDL window handle that is also the handle for MonoGame's Game.GameWindow instance. I don't see constructor or method in Eto for creating Forms (or any widget) with IntPtr handles. There doesn't seem to be an equivalent to the FromHandle method. How do I implement this in Eto in a way that is equivalent to how it was in winforms?

Since MonoGame is already handling the creation of the window, and much of my code operates around that, I am reluctant to create the MonoGame window using an Eto Control handler (as discussed here: https://github.com/picoe/Eto/issues/692#issue-204856552). That said, I'm open to suggestions for that as well.

**Comment by cwensley** [comment]
Hey @EclipsedButter,

Thanks for the question.  Usually to wrap native/platform specific windows and controls you'd reference the platform specific assembly e.g. Eto.WinForms, and use the helper methods e.g. `Eto.Forms.WinFormsHelpers.ToEtoWindow()`.  

That being said, it would be helpful to know what you want to do with the returned Window object?  The only things that would be wired up currently is the window size/location, and to be able to use it as a parent for dialogs/child forms.  There's no way to control the content of a wrapped native window with Eto.

**Comment by EclipsedButter** [comment]
My apologies, I should have clarified. I am migrating it to Eto so that I can run it on macOS, so I don't have access to `WinFormsHelpers`, only `MonoMac64Helpers`, and I don't have much experience with MonoMac compared to winforms.

In winforms, the `FromHandle` method let me define the a Form that used the same window that MonoGame draws on, which was of type `Microsoft.Xna.Framework.GameWindow`. Since I'm using MonoGame on Mac, the handle is for an SDL window. I had used this Form as the main form for the application, and had also used the same handle to create a few additional Forms in the same window. Since I don't think that much integration would be possible with Eto, I'm mainly looking for the best way to contain an SDL window as a child of a main Form object (or some similar structure; I just need to contain it). Since MonoGame exposes that window as an `IntPtr` handle, I'm looking for a way to contain it using that handle. It's fine if the content and control of the game window is self-contained and inaccessible to Eto itself.

**Reply by cwensley** [reply]
On macOS you'll have to get the IntPtr of the NSWindow itself in order to wrap it using something like this:

```c#
var nsWindowHandle = /* Some SDL method to get the native NSWindow handle from the SDL handle */
var nsWindow = MonoMac.ObjCRuntime.Runtime.GetNSObject<NSWindow>(nsWindowHandle);
var etoWindow = MonoMac64Helpers.ToEtoWindow(nsWindow);
```

Eto knows nothing about SDL so passing something other than an NSWindow won't do anything.

Hope this helps.

**Comment by ghost** [comment]
In Eto.Forms, there's no direct equivalent to Form.FromHandle for creating forms from an existing window handle. Here are a few options:

1. Platform-Specific Integration (Windows):
You can use P/Invoke to set the window handle in a custom Form for Windows.
Example:
`[DllImport("user32.dll", SetLastError = true)]
private static extern IntPtr SetParent(IntPtr hWndChild, IntPtr hWndNewParent);`

2. Eto Canvas for Rendering MonoGame:
Create a custom Canvas or Panel in Eto and render MonoGame content there.
Example:
`public class MonoGameCanvas : Eto.Forms.Canvas { ... }`

3. Use WinFormsHost (Windows Only):
On Windows, embed your MonoGame window in an Eto form using WinFormsHost.
Example:

var winFormsHost = new WinFormsHost();
These methods allow you to integrate the MonoGame window with Eto while maintaining cross-platform support.
